src/Compilers/CSharp/Portable/Symbols/Source/LocalFunctionSymbol.cs

```diff
@@ -13,55 +13,26 @@ namespace Microsoft.CodeAnalysis.CSharp.Symbols
 {
     internal sealed class LocalFunctionSymbol : MethodSymbol
     {
-        private sealed class ParametersAndDiagnostics
-        {
-            internal readonly ImmutableArray<ParameterSymbol> Parameters;
-            internal readonly bool IsVararg;
-            internal readonly ImmutableArray<Diagnostic> Diagnostics;
-
-            internal ParametersAndDiagnostics(ImmutableArray<ParameterSymbol> parameters, bool isVararg, ImmutableArray<Diagnostic> diagnostics)
-            {
-                Parameters = parameters;
-                IsVararg = isVararg;
-                Diagnostics = diagnostics;
-            }
-        }
-
-        private sealed class TypeParameterConstraintsAndDiagnostics
-        {
-            internal readonly ImmutableArray<TypeParameterConstraintClause> ConstraintClauses;
-            internal readonly ImmutableArray<Diagnostic> Diagnostics;
-
-            internal TypeParameterConstraintsAndDiagnostics(ImmutableArray<TypeParameterConstraintClause> constraintClauses, ImmutableArray<Diagnostic> diagnostics)
-            {
-                ConstraintClauses = constraintClauses;
-                Diagnostics = diagnostics;
-            }
-        }
-
-        private sealed class ReturnTypeAndDiagnostics
-        {
-            internal readonly TypeSymbol ReturnType;
-            internal readonly ImmutableArray<Diagnostic> Diagnostics;
-
-            internal ReturnTypeAndDiagnostics(TypeSymbol returnType, ImmutableArray<Diagnostic> diagnostics)
-            {
-                ReturnType = returnType;
-                Diagnostics = diagnostics;
-            }
-        }
 
         private readonly Binder _binder;
         private readonly LocalFunctionStatementSyntax _syntax;
         private readonly Symbol _containingSymbol;
         private readonly DeclarationModifiers _declarationModifiers;
         private readonly ImmutableArray<LocalFunctionTypeParameterSymbol> _typeParameters;
         private readonly RefKind _refKind;
-        private ParametersAndDiagnostics _lazyParametersAndDiagnostics;
-        private TypeParameterConstraintsAndDiagnostics _lazyTypeParameterConstraintsAndDiagnostics;
-        private ReturnTypeAndDiagnostics _lazyReturnTypeAndDiagnostics;
+
+        private ImmutableArray<ParameterSymbol> _lazyParameters;
+        private bool _lazyIsVarArg;
+        private ImmutableArray<TypeParameterConstraintClause> _lazyTypeParameterConstraints;
+        private TypeSymbol _lazyReturnType;
         private TypeSymbol _iteratorElementType;
-        private ImmutableArray<Diagnostic> _diagnostics;
+
+        private readonly DiagnosticBag _declarationDiagnostics;
+
+        // Lock for initializing lazy fields and registering their diagnostics
+        // Acquire this lock when initializing lazy objects to guarantee their declaration
+        // diagnostics get added to the store exactly once
+        private readonly object _lazyLock = new object();
```

Comment:
I was worried that `DiagnosticBag` would lock `this`. I didn't check if it did, though.